Error - No thumbnailer found for MIME type video/mp4

I started digitizing 8mm films and adding them as media objects in Gramps.  Using GrampsWeb when I browse the media object page (`/media/grampsID`), I receive the error message below in Docker.  In Gramps Desktop the file plays using the default app for that `mime/type` on the computer.  

It would be a nice enhancement to use the HTML `video` and `audio` tags to allow these files to be played on the Media Object page in GrampsWeb.   

It looks to me that the error below is complaining about the lack of a method to create a thumbnail for video mime-types.   GrampsWeb may need to have a generic thumbnail for audio files (https://commons.wikimedia.org/wiki/Category:Sound_file_icons) and video icons (https://commons.wikimedia.org/wiki/Category:Video_icons).  Presumably, with generic icons, it would allow the media objects listed in the Gallery Tab on People, Places, Events, etc.

`[2021-12-27 14:57:52 +0000] [14] [ERROR] Error handling request /api/media/ee9a340f6b8527ddf332bc9edd2/thumbnail/3000?jwt=tokenvalue&square=false
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 136, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 179, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2076, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/auth.py", line 45, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/webargs/core.py", line 450, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/flask_caching/__init__.py", line 475, in decorated_function
    rv = f(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/__init__.py", line 250, in get_thumbnail
    return handler.send_thumbnail(size=size, square=args["square"])
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/file.py", line 127, in send_thumbnail
    thumb = ThumbnailHandler(self.path_abs, self.mime)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/image.py", line 101, in __init__
    raise ValueError(
ValueError: No thumbnailer found for MIME type video/mp4.`

Thanks,
Bruce.

--- a/Dockerfile
+++ b/Dockerfile
@@ -3,8 +3,10 @@ FROM dmstraub/gramps:5.1.4
 WORKDIR /app
 ENV PYTHONPATH="${PYTHONPATH}:/usr/lib/python3/dist-packages"
 
-# install poppler (needed for PDF thumbnails)
-RUN apt-get update && apt-get install -y poppler-utils && rm -rf /var/lib/apt/lists/*
+# install poppler (needed for PDF thumbnails) and ffmpeg (needed for video thumbnails)
+RUN apt-get update && apt-get install -y \
+  poppler-utils ffmpeg libavcodec-extra \
+  && rm -rf /var/lib/apt/lists/*
 
 # set locale
 RUN localedef -i en_GB -c -f UTF-8 -A /usr/share/locale/locale.alias en_GB.UTF-8

--- a/gramps_webapi/api/image.py
+++ b/gramps_webapi/api/image.py
@@ -24,6 +24,7 @@ import io
 from pathlib import Path
 from typing import BinaryIO
 
+import ffmpeg
 from pdf2image import convert_from_path
 from PIL import Image, ImageOps
 
@@ -56,7 +57,11 @@ def image_square(image: Image) -> Image:
     """Crop an image to a centered square."""
     size = min(image.size)
     return ImageOps.fit(
-        image, (size, size), bleed=0.0, centering=(0.0, 0.5), method=Image.BICUBIC,
+        image,
+        (size, size),
+        bleed=0.0,
+        centering=(0.0, 0.5),
+        method=Image.BICUBIC,
     )
 
 
@@ -96,17 +101,24 @@ class ThumbnailHandler:
         self.mime_type = mime_type
         if self.mime_type.startswith("image/"):
             self.is_image = True
+            self.is_video = False
+        elif self.mime_type.startswith("video/"):
+            self.is_image = False
+            self.is_video = True
         else:
             if self.mime_type not in self.MIME_NO_IMAGE:
                 raise ValueError(
                     "No thumbnailer found for MIME type {}.".format(self.mime_type)
                 )
             self.is_image = False
+            self.is_video = False
 
     def get_image(self) -> Image:
         """Get a Pillow Image instance."""
         if self.mime_type == MIME_PDF:
             return self._get_image_pdf()
+        if self.is_video:
+            return self._get_image_video()
         return Image.open(self.path)
 
     def get_cropped(
@@ -129,6 +141,15 @@ class ThumbnailHandler:
         """Get a Pillow Image instance of the PDF's first page."""
         ims = convert_from_path(self.path, single_file=True, use_cropbox=True, dpi=100)
         return ims[0]
+
+    def _get_image_video(self) -> Image:
+        """Get a Pillow Image instance of the video's first frame."""
+        out, _ = (
+            ffmpeg.input(self.path, ss=0)
+            .output("pipe:", format="image2", pix_fmt="rgb24", vframes=1)
+            .run(capture_stdout=True, capture_stderr=True)
+        )
+        return Image.open(io.BytesIO(out))
 
     def get_thumbnail(
         self, size: int, square: bool = False, fmt: str = "JPEG"

--- a/setup.py
+++ b/setup.py
@@ -45,8 +45,6 @@ REQUIREMENTS = [
     "jsonschema",
 ]
 
-EXTRA_REQUIREMENTS = ["pycairo", "PyGObject"]
-
 setup(
     author="Gramps Development Team",
     url="https://github.com/gramps-project/web-api",
@@ -54,7 +52,6 @@ setup(
     description="A RESTful web API for the Gramps genealogical database.",
     license="AGPL v3 or greater",
     install_requires=REQUIREMENTS,
-    extra_requires=EXTRA_REQUIREMENTS,
     long_description=README,
     long_description_content_type="text/markdown",
     include_package_data=True,

--- a/setup.py
+++ b/setup.py
@@ -43,6 +43,7 @@ REQUIREMENTS = [
     "bleach",
     "whoosh",
     "jsonschema",
+    "ffmpeg-python",
 ]
 
 setup(

